### PR TITLE
Split the allow list and increase the warning level for forever BC breaking changes

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72479
* __->__ #72474

Differential Revision: [D34057168](https://our.internmc.facebook.com/intern/diff/D34057168)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 1]
```diff
@@ -26,110 +26,94 @@
...
-    ("aten::_native_multi_head_self_attention", datetime.date(9999, 1, 1)),
 ]
 
-ALLOW_LIST_COMPILED = [
-    (
-        re.compile(item[0]),
-        item[1],
-        re.compile(item[2]) if len(item) > 2 else None,
-    ) for item in ALLOW_LIST if item[1] >= datetime.date.today()
+MAX_ALLOWED_PERIOD = datetime.timedelta(days=365)
```
Comment (tugsbayasgalan) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:43]: Rn this is very high until we figure out what we do with the operators that have long future dates

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 2]
```diff
@@ -26,110 +26,94 @@
...
+
+# WARNING: The operators that are included in this list is extremely unsafe
+# because they would bypass all BC/FC checks for indefinite amount of time.
+# This could cause many severe production issues. Please proceed with extreme caution,
+# if you must add an entry to this list.
+ALLOW_LIST_WITHOUT_DEADLINE = [
+    "c10_experimental",
+    "static",
```
Comment (gmagogsfm) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:51]: There used to be an `Internal` comment line here

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 3]
```diff
@@ -162,9 +146,12 @@ def check_bc(existing_schemas):
     is_bc = True
     broken_ops = []
     for existing_schema in existing_schemas:
-        if allow_listed(existing_schema):
+        if allow_listed_with_deadline(existing_schema):
             print("schema: ", str(existing_schema), " found on allowlist, skipping")
             continue
+        if allow_listed_without_deadline(existing_schema):
+            print("schema: {} is found in forever BC breaking list. This is very dangerous".format(str(existing_schema)))
```
Comment (gmagogsfm) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:153]: Maybe rephrase it to be something like:
{} is in allowlist for BC-breaking evolution without deadline. This is dangerous, do not use unless you are sure there will not be downstream consequences.

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 4]
```diff
@@ -199,7 +186,7 @@ def check_fc(existing_schemas):
     is_fc = True
     broken_ops = []
     for existing_schema in existing_schemas:
-        if allow_listed(existing_schema):
+        if allow_listed_with_deadline(existing_schema):
```
Comment (gmagogsfm) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:189]: I thought we were going to have a separate list for FC breaking allowlist? And why isn't the list without deadline checked here?

Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:189]: I was gonna do that in a separate PR. In thi PR, I tried not to touch the fc part of the test. But yeah, i think it makes sense to add the one without deadline as well.

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 3 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:153]: sounds good.

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 5]
```diff
@@ -26,110 +26,94 @@
...
-    ("aten::slow_conv3d_forward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv3d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_dilated2d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_dilated3d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose2d", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose2d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose3d", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose3d_backward", datetime.date(2022, 1, 31)),
+ALLOW_LIST_WITH_DEADLINE = [
     ("aten::_svd_helper", datetime.date(2022, 3, 31)),
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:30]: I think all of these can be removed?

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 6]
```diff
@@ -26,110 +26,94 @@
...
-    ("aten::_slow_conv2d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv3d_forward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv3d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_dilated2d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_dilated3d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose2d", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose2d_backward", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose3d", datetime.date(2022, 1, 31)),
-    ("aten::slow_conv_transpose3d_backward", datetime.date(2022, 1, 31)),
+ALLOW_LIST_WITH_DEADLINE = [
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:29]: Let's extend the comment here to describe what this list is for and how to use it properly (with a good example and a bad example, maybe?)

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 7]
```diff
@@ -26,110 +26,94 @@
...
 
-ALLOW_LIST_COMPILED = [
-    (
-        re.compile(item[0]),
-        item[1],
-        re.compile(item[2]) if len(item) > 2 else None,
-    ) for item in ALLOW_LIST if item[1] >= datetime.date.today()
+MAX_ALLOWED_PERIOD = datetime.timedelta(days=365)
+
+# WARNING: The operators that are included in this list is extremely unsafe
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:45]: This is a little ungrammatical. Maybe something more like: "Operators included in this list indefinitely bypass all BC and FC schema checks. This is almost certainly NOT what you want to do. See note above."?

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 1 above)
Comment (mruberry) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:43]: I would make this one month and put it above the first list which it applies to, then include in the comment there what this does

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 6 above)
Comment (mruberry) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:29]: "TEMPORARY_ALLOW_LIST"?

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 8]
```diff
@@ -26,110 +26,94 @@
...
+# because they would bypass all BC/FC checks for indefinite amount of time.
+# This could cause many severe production issues. Please proceed with extreme caution,
+# if you must add an entry to this list.
+ALLOW_LIST_WITHOUT_DEADLINE = [
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:49]: "INDEFINITE_ALLOW_LIST"?

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 9]
```diff
@@ -26,110 +26,94 @@
...
-    for item in ALLOW_LIST_COMPILED:
+def compile_allow_list_with_deadline():
+    output = []
+    for item in ALLOW_LIST_WITH_DEADLINE:
+        deadline = item[1]
+        today = datetime.date.today()
+        if deadline > today and deadline - today < MAX_ALLOWED_PERIOD:
+            output.append((re.compile(item[0]), deadline, re.compile(item[2]) if len(item) > 2 else None))
+        if deadline - today >= MAX_ALLOWED_PERIOD:
+            print("{} will be BC broken for too long. We only allow {} days"
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:95]: This is misleading. The operators in question are not necessarily "broken" for this period, they are just scheming schema BC and FC checks.

This warning should also provide guidance about how the test works and what is an appropriate period to set (not just what is allowed). For example, "Operator foo will skip BC and FC schema checks for X days, but only Y days of skipping the checks are permitted. It's recommended that the skip date be chosen only to permit the PR to be merged. Once the PR is merged for a couple days the operator no longer needs to skip these checks."

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 3 above)
Comment (mruberry) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:153]: We should be clear this is skipping BC and FC schema checks, but I think we would need to add more details about exactly how this is "dangerous" and when, if ever, it's OK to do this. If we're not sure then having someone contact us is probably the appropriate recommendation

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 10]
```diff
@@ -232,7 +219,7 @@ def check_fc(existing_schemas):
     if is_fc:
         print("Found forward compatible schemas for all existing schemas")
     else:
-        warnings.warn(
+        print(
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:222]: Should this be upgraded to a failure? It's hard to spot prints and warnings.

The error message here is a little goofy. The first sentence is good, but what does it mean to "contact the PyTorch team"? The operators are also listed as "broken" but they are not necessarily broken, they just a potential forward compatibility schema mismatch. We should try to be as clear as possible in our language so developers can understand the issues our test suite presents to them.

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 2 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:44]: done

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 5 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py]: Sounds good

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 8 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py]: Changed

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 7 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py]: Changed

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 10 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py]: Changed

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 11]
```diff
@@ -26,111 +32,84 @@
...
-    ) for item in ALLOW_LIST if item[1] >= datetime.date.today()
-]
+def compile_temp_allow_list():
+    output = []
+    for item in TEMPORARY_ALLOW_LIST:
+        deadline = item[1]
+        today = datetime.date.today()
+        interval = deadline - today
+        if deadline > today and deadline - today < MAX_ALLOWED_PERIOD:
+            output.append((re.compile(item[0]), deadline, re.compile(item[2]) if len(item) > 2 else None))
```
Comment (gmagogsfm) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:87]: What is `item[2]`?

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 12]
```diff
@@ -26,111 +32,84 @@
...
         if item[0].search(str(schema)):
             if len(item) > 2 and item[2] is not None:
                 # if arguments regex is present, use it
                 return bool(item[2].search(str(schema)))
             return True
     return False
 
+def indefinite_allow_listed(schema):
+    for item in INDEFINITE_ALLOW_LIST:
+        if re.compile(item).search(str(schema)):
```
Comment (gmagogsfm) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:110]: Maybe we should pre-compile re in `INDEFINITE_ALLOW_LIST` too?

Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:110]: Why?

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py (hunk 11 above)
Comment (tugsbayasgalan) [reply test/forward_backward_compatibility/check_forward_backward_compatibility.py:87]: This was here before. (https://github.com/pytorch/pytorch/pull/72474/files#diff-236fbde71e59cb1597cac177a83e49fb62b30770eec55c4e7a0f2650b9eb6203L25) And this is pretty much a copy of what we had before. But now I think about it, it is probably ok to delete it

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).

File: test/forward_backward_compatibility/check_forward_backward_compatibility.py [hunk 13]
```diff
@@ -8,127 +8,105 @@
 import torch
 from torch._C import parse_schema
 
+# Maximum days we will allow to introduce BC
+# breaking change into operators.
+MAX_ALLOWED_PERIOD = datetime.timedelta(days=30)
 
 # The date specifies how long the allowlist exclusion should apply to.
+# You should pick a date that is far enough in the future that you
```
Comment (mruberry) [inline test/forward_backward_compatibility/check_forward_backward_compatibility.py:16]: Nice comment; this is what I was mistakenly looking for in the next PR

**Review by gmagogsfm — APPROVED** [review]

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34057168).
